### PR TITLE
LibraryControl: added "TriggerItem" control object / code, numark mixtrack 3 script now using TriggerItem

### PR DESCRIPTION
**Pull request by shuda**
**TriggerItem(double v)** works similarly as the **GotoItem** (technically a wrapper for gotoitem)
Input param exposes more user/script control and has some auto functionality.

**If value equal -1:**
GotoItem called internally without any changes.

**If value equal 0:**
Function will check if any valid toggleable item is selected on the sidebar widget and toggle selected item.
Otherwise it calls GotoItem

**If value equal 1-5:**
target library focus method, convert value to @FocusWidget enum and set library focus (forced) before calling gotoitem

### CODE REVIEW

File: src/library/librarycontrol.cpp [hunk 1]
```diff
@@ -976,6 +990,36 @@ void LibraryControl::slotGoToItem(double v) {
...
+            m_pSidebarWidget->toggleSelectedItem();
+        } else
+            slotGoToItem(1);
+        return;
+    }
+
+    // todo: update if focuswidget enum changes
+    // enum overrun
+    if (triggerItemMethod > 5)
+        return;
```
Comment (ronso0) [inline src/library/librarycontrol.cpp:1015]:
```suggestion
    if (triggerItemMethod > static_cast<int>(FocusWidget::Count)) {
        return;
    }
```
(actually it should be limited to the library widgets which are always available (in official skins), but `setLibraryFocus()` already takes care of that

**Comment by ronso0** [comment]
Thank you for this PR.

Please explain the benefits, ideally with distinct use cases.

**Review by ronso0 — CHANGES_REQUESTED** [review]

File: src/library/librarycontrol.cpp [hunk 2]
```diff
@@ -976,6 +990,36 @@ void LibraryControl::slotGoToItem(double v) {
...
+    if (triggerItemMethod == -1) {
+        slotGoToItem(1);
+        return;
+    }
+
+    // check and try toggle items selected on sidebar first if possible
+    if (triggerItemMethod == 0) {
+        if (!m_pSidebarWidget->isLeafNodeSelected()) {
+            setLibraryFocus(FocusWidget::Sidebar);
+            m_pSidebarWidget->toggleSelectedItem();
```
[the remark this answers is not in the document]
Comment (shuda) [reply src/library/librarycontrol.cpp:1006]: GotoItem is bugging when shift + browse button used on numark mixtrack pro 3.
Sometimes focus just go off to other widget and browse button (toggling on sidebar) dont do anything.
Tried to fix that via controller scripting but in the end gotoitem gives focus (or something else takes the focus) to other widget.

Thought it will be a clever idea to have a function with exposed param and can be manipulated/controlled by users via scripting instead of just do things automatically in c++.
tldr: gotoitem but give basic control at least, so users can manipulate functionality later based on their own requirements and current focused widget (if accessible via scripting)

[the remark this answers is not in the document]
Comment (ronso0) [reply src/library/librarycontrol.cpp:1006]: Hmm, to me it seems `TriggerItem` is supposed to work around the limitations of the MT3 mapping, uses a questionable auto-behaviour with `0` (and enforces that on all MT3 users).

IMO the library navigation on the Mixtrack 3 with the Browse knob is sub-optimal:
* it uses deprecated `[Playlist]` controls to navigate the sidebar and the tracks table (and I don't understand why you care about focus when using these controls?)
* no focus controls, i.e. you can't go to the searchbar with the controller and select a saved query (new in Mixxx 2.4)

Ccorrect my if I'm wrong, all this can be achieved with the current ("new") library controls:
* `[Library],MoveFocus` (like Tab / Shift+Tab)
* `[Library],MoveVertical` (like Up/Down keys) to move the selection in the focused pane
* if you need custom behaviour, e.g. jump only between the sidebar and tracks table on Browse press, use `[Library],focused_widget` (read/write)

We spent a lot of time discussing these controls and how they're supposed to work in various use cases. Please describe which use case we missed that justifies adding yet another control.

[the remark this answers is not in the document]
Comment (shuda) [reply src/library/librarycontrol.cpp:1006]: ohh wait, `[Library],focused_widget` is readable?
ohm i made a big mistake... 
i've read wrong documentation for the entire time, 2.3 controls doc was bookmarked instead of 2.4..... meh...

now i see why my changes makes no sense..
focused_widget didnt exists in 2.3, thought controller scripts not allowed to change widget focus.. but.. i was simply blind :)

you have right, no new control is needed. browse button bug can be fixed entirely in the controller script.
my bad.

[the remark this answers is not in the document]
Comment (ronso0) [reply src/library/librarycontrol.cpp:1006]: Alright :wink: 

Nevertheless, the MT3 mapping should be updated with the new(er) controls so it makes all new 2.4 features accessible. Do you have motivation to do that?
The MT3 layout is generic so I think this would work well (and users updating to Mixxx 2.4 will adapt to the new fuctions quickly IMO):
* **Browse** turn: `[Library],MoveVertical`
* **Shift** + **Browse** turn: move focus
* **Browse** press: GoToItem
* **Shift**+**Browse** press: maximize library

What do you think?

[the remark this answers is not in the document]
Comment (ronso0) [reply src/library/librarycontrol.cpp:1006]: So we can close this PR?
Sorry you spent time on this...

[the remark this answers is not in the document]
Comment (shuda) [reply src/library/librarycontrol.cpp:1006]: i closed the pr :)
and well.. yes.. but still.. time was valuable..
im coming from gamedev/c++ programming, dj app and the entire mixxx code is relatively new for me..
at least i got to know mixxx's source code / system's better

[the remark this answers is not in the document]
Comment (ronso0) [reply src/library/librarycontrol.cpp:1006]: What do you think about the update described I described above? https://github.com/mixxxdj/mixxx/pull/12497#discussion_r1440008191

[the remark this answers is not in the document]
Comment (shuda) [reply src/library/librarycontrol.cpp:1006]: sure, gonna do that.. tomorrow or end of this week... (or sooner if i have time.. just have to prepare a steam patch @ work :'( :D )
